Ls/standardize compound unit for samples/144

lshandross opened the pull request:
Adds ability to subset provided samples, implemented by a map over groups defined by the columns making up the compound task id set and sampling the output type ids of the requested number of predictions to return.

elray1 commented on R/linear_pool_sample.R:
```diff
@@ -21,18 +21,17 @@ linear_pool_sample <- function(model_out_tbl, weights = NULL,
                                task_id_cols = NULL,
+                               comp_unit_cols = NULL,
                                n_output_samples = NULL) {
 
```
Question about this function: are we doing something in here to ensure that `output_type_id`s from different models are distinct in the ensemble?  It may be in there, I just didn't see it.

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -453,13 +453,130 @@ test_that("samples only collected and re-indexed for simplest case", {
+    dplyr::mutate(component_model = ifelse(component_model == letters[5], models_to_resample, component_model)) |>
+    dplyr::arrange(dplyr::across(c("target", "location"))) |>
+    dplyr::tibble()
+  subset_actual <- sample_outputs |>
+    dplyr::mutate(component_model = .data[["model_id"]], .before = 1) |>
+    linear_pool(
+      task_id_cols = c("target_date", "target", "horizon", "location"),
+      comp_unit_cols = c("target", "location"),
+      n_output_samples = 5
+    ) |>
+    dplyr::group_by(dplyr::across(dplyr::all_of(c("target", "location", "output_type_id", "component_model")))) |>
+    dplyr::summarize(distinct_ids = dplyr::n()) |>
+    dplyr::ungroup() |>
+    dplyr::select(-"output_type_id")
+  expect_equal(subset_actual, subset_expected)
```
Also, do we check that output type ids from different component models are distinct in the ensemble output?  I see you are doing something with a variable called `distinct_ids` here, but I'm not totally clear on what that variable is doing.

elray1 replied:
Overall, it would be helpful to me to enumerate (in a comment) the different things we expect to see in the output from this function and how we're checking it.

lshandross replied:
Basically, this test is checking that for requested `n_output_samples` that don't divide evenly among models, the remainder is correctly split among the models (i.e. all the remainder samples aren't requested from one single model). I probably should actually do `n_output_samples = 6` to better test this code.

The `distinct_ids` is used to show if the dependencies are properly being accounted for, i.e. for this example in the test the same forecast (generated by output type id) is being drawn for horizons 0 and 1 for the same compound task set values. I don't have an explicit test that the output type ids from different component models are distinct in the ensemble output but I will add one.

elray1 replied:
restating the second paragraph above: within each group defined by values of model id and comp_taskid_set variables, each output_type_id shows up for all values of variables that are not in the comp_taskid_set (i.e., for all horizons if we get trajectories)

lshandross replied:
Would it be useful to add an in-line comment within the testing file to explain this?

elray1 replied:
Yeah, I think that would be good! Thanks.

elray1 replied:
I'm marking this comment as resolved.  there are a couple of things in here that are repeated in other comments.

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -453,13 +453,130 @@ test_that("samples only collected and re-indexed for simplest case", {
+test_that("ensemble of samples correctly drawn for compound task ID sets", {
+  sample_outputs <- expand.grid(stringsAsFactors = FALSE,
+                                model_id = letters[1:4],
+                                location = c("222", "888"),
```
Per comments on the code above, let's add a test where different models submit for different subsets of locations.

lshandross replied:
Implemented! (https://github.com/hubverse-org/hubEnsembles/pull/147/commits/2c89864f3e16eb47b5e54a3749b5292251ee4a20)

elray1 reviewed (changes requested):
This is a good start on a challenging problem!  I have asked some questions throughout.

lshandross commented on R/linear_pool_sample.R:
```diff
@@ -42,17 +41,55 @@ linear_pool_sample <- function(model_out_tbl, weights = NULL,
+    }
+
+    # deal with n_output_samples not divisible evenly among component models
+    actual_output_samples <- samples_per_combo |>
+      dplyr::group_by(dplyr::across(dplyr::all_of(task_id_cols))) |>
+      dplyr::summarize(target_samples = sum(.data[["target_samples"]])) |>
+      dplyr::pull(.data[["target_samples"]]) |>
+      unique()
```
As mentioned in my previous comment, a situation like this will not arise with the code as it currently is because it will throw a validation error when not all component models provide the same number of sample forecasts for all possible combinations of task ID variables.

I had previously played around by doing the group by operation on `comp_unit_cols` instead but found that it didn't work with the way that the function is currently coded. This is because `n_output_samples` gives the number of sample forecasts to output for each unique combination of task ID variables, NOT compound task ID sets. I think it makes sense to keep this definition, because while forecasts for different values within a dependent task are related, they are still distinct forecasts (e.g. horizon 0 forecast made today for MA flu hospitalizations is distinct from that of the horizon 1 forecast)

lshandross replied:
Code logic changed to group by `comp_unit_cols` in commit https://github.com/hubverse-org/hubEnsembles/pull/147/commits/fb2312a6c4aa6ff79f54f7fad0bd8d7947e05ba5 to be more intuitive

lshandross replied:
Remainder samples now are calculated on a by–compound task ID set basis to avoid the bug mentioned above with changes implemented in commit https://github.com/hubverse-org/hubEnsembles/pull/147/commits/aaf04c9628f6cd8cf89995416a968df75cca02af

lshandross commented on R/linear_pool_sample.R:
```diff
@@ -42,17 +41,55 @@ linear_pool_sample <- function(model_out_tbl, weights = NULL,
+    model_out_tbl <- split_comp_units |>
+      purrr::map(.f = function(split_outputs) {
+        current_comp_unit <- split_outputs |>
+          dplyr::distinct(dplyr::across(dplyr::all_of(comp_unit_cols)), .keep_all = TRUE) |>
```
It gives us a way to get/store all the information about the task ID column values for the current iteration so that we can filter out the correct row from the `samples_per_combo` data frame, which gives us how many samples (labeled by unique indices) we're interested in drawing from the current `split_outputs` data frame.

The `keep_all` argument is just a logical asking whether to keep every column when performing the filter for distinct values.

lshandross commented on R/linear_pool_sample.R:
```diff
@@ -42,17 +41,55 @@ linear_pool_sample <- function(model_out_tbl, weights = NULL,
+          dplyr::left_join(samples_per_combo, by = c("model_id", task_id_cols))
+        provided_indices <- unique(split_outputs$output_type_id)
+
+        # check for same number of output type ids per modeling unit
```
Left over comment that's no longer relevant to the current code — will delete

lshandross commented on R/validate_ensemble_inputs.R:
```diff
@@ -119,6 +107,57 @@ validate_ensemble_inputs <- function(model_out_tbl, weights = NULL,
+#' @return no return value
+#' @noRd
+
+validate_weights <- function(model_out_cols, weights = NULL, weights_col_name = "weight") {
```
Filed as issue #149

lshandross replied:
Ended up realizing I could implement this validation by changing an out-dated one in `validate_sample_inputs()`, so I made that change in https://github.com/hubverse-org/hubEnsembles/pull/147/commits/32d3988d6b51be1e0bb5e3db7ba114f7f90012da

elray1 commented on R/linear_pool_sample.R:
```diff
@@ -42,15 +35,77 @@ linear_pool_sample <- function(model_out_tbl, weights = NULL,
+        fill = list(provided_samples = 0)
+      ) |>
+      dplyr::left_join(weights, weight_by_cols) |>
+      dplyr::mutate(target_samples = ifelse(
+        .data[["provided_samples"]] <= 0,
+        0,
+        floor(.data[[weights_col_name]] * n_output_samples)
+      ))
```
It seems like we probably need to re-normalize the model weights within groups defined by `compound_taskid_set`.  Borrowing an example from the last round of review, suppose `compound_taskid_set` includes `"location"`, and we have three equally-weighted models that each submitted 100 samples, but one of those models does not submit predictions for one location.  Then this calculation would result in only 66 ensemble samples.  I think this can be remedied by updating the weights using code something like this (untested):
```suggestion
      dplyr::group_by(dplyr::across(dplyr::all_of(compound_taskid_set))) |>
      dplyr::mutate(
        effective_weight = as.integer(.data[["provided_samples"]] > 0) * .data[[weights_col_name]],
        effective_weight = .data[["effective_weight"]] / sum(.data[["effective_weight"]]),
        target_samples = floor(.data[["effective_weight"]] * n_output_samples)
      ) |>
      dplyr::ungroup()
```

elray1 commented on R/linear_pool_sample.R:
```diff
@@ -42,15 +35,77 @@ linear_pool_sample <- function(model_out_tbl, weights = NULL,
+    # deal with n_output_samples not divisible evenly among component models
+    samples_per_combo <- samples_per_combo |>
+      purrr::map(.f = function(split_per_combo) {
+        valid_models <- split_per_combo$model_id[split_per_combo$provided_samples > 0]
+        split_per_combo[[weights_col_name]] <-
+          ifelse(split_per_combo$provided_samples == 0, 0, 1 / length(valid_models))
+        split_per_combo$target_samples <-
+          ifelse(split_per_combo$target_samples == 0, 0, floor(split_per_combo[[weights_col_name]] * n_output_samples))
+
+        actual_output_samples <- split_per_combo |>
+          dplyr::group_by(dplyr::across(dplyr::all_of(compound_taskid_set))) |>
+          dplyr::summarize(output_samples = sum(.data[["target_samples"]])) |>
+          dplyr::pull(.data[["output_samples"]]) |>
+          unique()
+        remainder_samples <- n_output_samples - actual_output_samples
+
+        models_to_resample <- sample(x = valid_models, size = remainder_samples)
+        split_per_combo |>
+          dplyr::mutate(target_samples = ifelse(
+            model_id %in% models_to_resample, .data[["target_samples"]] + 1, .data[["target_samples"]]
+          ))
```
this is a bit longer than I would usually specify for an anonymous function: typically, I'll aim to refactor an anonymous function into a separate named function when it's longer than ~2-3 lines

elray1 commented on R/linear_pool_sample.R:
```diff
@@ -42,15 +35,77 @@ linear_pool_sample <- function(model_out_tbl, weights = NULL,
+    # deal with n_output_samples not divisible evenly among component models
+    samples_per_combo <- samples_per_combo |>
+      purrr::map(.f = function(split_per_combo) {
+        valid_models <- split_per_combo$model_id[split_per_combo$provided_samples > 0]
+        split_per_combo[[weights_col_name]] <-
+          ifelse(split_per_combo$provided_samples == 0, 0, 1 / length(valid_models))
+        split_per_combo$target_samples <-
+          ifelse(split_per_combo$target_samples == 0, 0, floor(split_per_combo[[weights_col_name]] * n_output_samples))
```
I see now that you are addressing the comment I made above here.  I guess I'd say though that it seems like we should only need one of these calculations of the `target_samples`, either above or here.  If you'd prefer, I'd imagine it would be safe to remove the last mutate block that I commented on above, just keeping the computation of `target_samples` that you do here?

elray1 replied:
Or alternatively, I'd imagine that you could keep the computation above and remove this, and then this anonymous function becomes shorter and you could ignore my comment below about refactoring it.

elray1 commented on R/linear_pool_sample.R:
```diff
@@ -42,15 +35,77 @@ linear_pool_sample <- function(model_out_tbl, weights = NULL,
+        split_per_combo$target_samples <-
+          ifelse(split_per_combo$target_samples == 0, 0, floor(split_per_combo[[weights_col_name]] * n_output_samples))
+
+        actual_output_samples <- split_per_combo |>
+          dplyr::group_by(dplyr::across(dplyr::all_of(compound_taskid_set))) |>
+          dplyr::summarize(output_samples = sum(.data[["target_samples"]])) |>
+          dplyr::pull(.data[["output_samples"]]) |>
+          unique()
```
Since we are operating on `split_per_combo`, which was split based on the `compound_taskid_set`, I think this simpler computation is available:
```suggestion
        actual_output_samples <- sum(split_per_combo$target_samples)
```

elray1 commented on R/linear_pool_sample.R:
```diff
@@ -42,15 +35,77 @@ linear_pool_sample <- function(model_out_tbl, weights = NULL,
+  }
+
+  if (!is.null(n_output_samples)) {
+    samples_per_combo <- model_out_tbl |>
```
I would be inclined to pull out all of the logic from lines 46 - 88 into a separate function called something like `get_samples_per_combo`.  This is a fairly lengthy and conceptually distinct chunk of code.

elray1 reviewed (changes requested):
This is getting there!  I had several comments, but only one thing that I think is a substantive change or question.  The others are more minor cleanup.

I ran out of mental energy to read through the unit tests.  I'll return to those on Monday.

elray1 commented:
Here are some thoughts on test cases.  Several of these are already done, as indicated by a ✅ .

- ✅ In the simplest case (all models have the same number of samples, no weights, no limit on number of ensemble samples), we simply collect the individual model samples and update their indices to be different for different models.

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -411,6 +418,129 @@ test_that("(weighted) quantiles correctly calculated - lognormal family", {
 })
 
 
+test_that("Not all component models forecasting for the same set of dependent task values throws an error", {
+  sample_outputs <- expand.grid(stringsAsFactors = FALSE,
```
```suggestion
test_that("Not all component models forecasting for the same set of dependent task values throws an error", {
  # There are four models, "a", "b", "c", and "d".
  # The first three have samples for horizons 0 and 1, while model "d" has samples for only horizon 1.
  # The compound task id set doesn't include horizon, so the samples are trajectories over time.
  # We expect to see an error in this case, as we can't combine predictions from models with different
  # subsets of values for variables outside of the compount_taskid_set.
  sample_outputs <- expand.grid(stringsAsFactors = FALSE,
```

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -411,6 +418,129 @@ test_that("(weighted) quantiles correctly calculated - lognormal family", {
+})
+
+
+test_that("Unequal samples across component models for unique of compound task ID set vars throws an error", {
+  sample_outputs <- expand.grid(stringsAsFactors = FALSE,
```
```suggestion
test_that("Unequal samples across component models for unique of compound task ID set vars throws an error", {
  # there are four models, "a", "b", "c", and "d".
  # The first three models each submit 3 samples, while model "d" submits only 1 sample.
  # We expect an error in this situation, because our methods currently do not support it.
  sample_outputs <- expand.grid(stringsAsFactors = FALSE,
```

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -411,6 +418,129 @@ test_that("(weighted) quantiles correctly calculated - lognormal family", {
+  sample_outputs |>
+    dplyr::mutate(horizon = 0, value = 0.75 * value) |>
+    dplyr::bind_rows(sample_outputs) |>
+    dplyr::filter(model_id %in% letters[1:3] | output_type_id == 1) |>
```
To make this test more specific to the "within each compound unit", we could add a condition here to only do this filter within one location:
```suggestion
    dplyr::filter(model_id %in% letters[1:3] | (output_type_id == 1 & location == "222")) |>
```

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -453,13 +583,211 @@ test_that("samples only collected and re-indexed for simplest case", {
 })
 
 
+test_that("remainder samples are properly distributed when component models don't all forecast for every location", {
+  sample_outputs <- expand.grid(stringsAsFactors = FALSE,
```
Can we add a comment here, similar to examples I added above, describing the content of the test?

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -411,6 +418,129 @@ test_that("(weighted) quantiles correctly calculated - lognormal family", {
 
 
+test_that("Not all component models forecasting for the same set of dependent task values throws an error", {
+  sample_outputs <- expand.grid(stringsAsFactors = FALSE,
```
It may be too late now for this to be a useful suggestion, but since we use this example repeatedly, it could be helpful to pull its definition out into a helper function.

elray1 replied:
I do think this would be useful to do: it would make this file shorter, and it would make it easier to find the difference in the setup for each test case.  For example, in this test we would take a complete, valid submission with both horizons and then "break" it by filtering it on the combination of `model_id` and `horizon`.

One way to achieve this could be to create a file called `tests/testthat/helper.R` or `tests/testthat/helper-test_data.R`, or something else like that, and within that define a function like `create_test_sample_outputs()` that creates a generally useful/usable example of `sample_outputs`.  Then in this function, you can call `sample_outputs <- create_test_sample_outputs()` and modify it as needed.  For example, you might have something like:
```r
sample_outputs <- create_test_sample_outputs() |>
  dplyr::filter(model_id %in% letters[1:3] | horizon == 1)
```

This setup is described in the R packages book in these two sections:
 - [helper files](https://r-pkgs.org/testing-design.html#testthat-helper-files)
 - [test fixtures](https://r-pkgs.org/testing-advanced.html#test-fixtures)

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -453,13 +583,211 @@ test_that("samples only collected and re-indexed for simplest case", {
+    dplyr::group_by(dplyr::across(dplyr::all_of(
+      c("location", "target", "target_date", "output_type_id", "component_model")
+    ))) |>
+    dplyr::summarize(distinct_ids = dplyr::n()) |>
```
Can you add some comments explaining what `distinct_ids` is measuring and why in a correct test result, we expect that its value is 1?

lshandross replied:
I'm not sure the best way to express this succinctly for an inline comment, but the `distinct_ids` variable is used to check whether the sampling among the component models is performed correctly for the non-compound task id set variables (the ones with dependency). In this case, we have trajectory samples only for `horizon`, but there is only `horizon = 1` in the toy data so we expect `distinct_ids = 1`. However, if the toy data has `horizon = 0:1`, then we should have `distinct_ids = 2`, which is the case in a later test.

Perhaps a better name would be `joint_across_samples`? The original name was created based on the fact that I was calculating the value using the number of distinct output type ids

elray1 replied:
I'm still having a hard time thinking about this test.  What would a failing `linear_pool` output look like?

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -453,13 +583,211 @@ test_that("samples only collected and re-indexed for simplest case", {
+    dplyr::group_by(dplyr::across(dplyr::all_of(
+      c("location", "target", "target_date", "output_type_id", "component_model")
+    ))) |>
+    dplyr::summarize(distinct_ids = dplyr::n()) |>
```
Similar question as on the previous test: I'm not exactly sure what `distinct_ids` is measuring.

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -453,13 +583,211 @@ test_that("samples only collected and re-indexed for simplest case", {
+    dplyr::mutate(horizon = 0, value = 0.75 * value) |>
+    dplyr::bind_rows(sample_outputs)
+
+  # All compound units have unique ids which are shared across dependent task columns
```
What do you mean by "All compound units have unique ids" -- do you mean `output_type_id`s?

lshandross replied:
Yes — I've updated the in-line comments to be clearer

lshandross commented on R/linear_pool.R:
```diff
@@ -58,6 +62,7 @@ linear_pool <- function(model_out_tbl, weights = NULL,
                         weights_col_name = "weight",
                         model_id = "hub-ensemble",
                         task_id_cols = NULL,
+                        compound_taskid_set = NULL,
```
Originally, I had agreed with the decision to remove a default value, but upon making the change I realized that now users will have to specify the compound_taskid_set even if it is not being used to calculate the linear pool they're requesting (e.g., a linear pool of non-sample output type forecasts). Is this behavior that we want? Or would we prefer to set a default value of NA (which would also signify the compound_taskid_set is not relevant) and write a check that throws an error if we actually do need a value for the compound_taskid_set? (I'm leaning towards the latter)

lshandross replied:
Went ahead and implemented the second option

lshandross commented on R/validate_ensemble_inputs.R:
```diff
@@ -119,6 +98,127 @@ validate_ensemble_inputs <- function(model_out_tbl, weights = NULL,
+}
+
+
+#' Perform validations on the compound task ID set used to calculate an ensemble of
+#' component model outputs for each combination of model task, output type,
+#' and output type id for the sample output type
```
1. I rewrote the function description to get rid of that part, which was left over from copying part of another function's docs that I shouldn't have.
2. Made this change
3. This case is not caught but based on your later suggestions I've added a check that the provided `compound_taskid_set` is compatible with the model outputs by checking that for each individual model, the output type ID levels are shared across all levels of non-compound task ID set variables

elray1 commented on R/linear_pool.R:
```diff
@@ -9,6 +9,15 @@
+#'   all task id variables, while equality to `task_id_cols` means that the samples
+#'   are from separate univariate distributions for each individual prediction task.
+#'   NA means the compound_taskid_set is not relevant for the current modeling task.
+#'   Defaults to NA. Derived task ids must be included if ANY of the task ids their
```
Changing "ANY" to "all" below:
```suggestion
#'   Defaults to NA. Derived task ids must be included if all of the task ids their
```
Here are some examples for the reference_date/horizon/target_date trio, regarding horizon as derived:
 - hub collects independent predictions for each horizon/target_date: reference_date and target_date go into the compound_taskid_set, so horizon must also go there
 - hub collects trajectories across horizon/target_date for each referene_date: compound_taskid_set includes reference_date, but not horizon or target_date

lshandross replied:
I had actually done some testing about these cases and found that for the second, `target_date` actually did have to go into the compound task id set, hence "any" instead of "all" in the documentation. Maybe we can talk about this synchronously to clear it up further

lshandross replied:
Looking further into this, I discovered a bug with the validations code for sample outputs. I've summarized it and my fixes in a later comment

elray1 commented on R/linear_pool.R:
```diff
@@ -9,6 +9,15 @@
+#'   NA means the compound_taskid_set is not relevant for the current modeling task.
+#'   Defaults to NA. Derived task ids must be included if ANY of the task ids their
+#'   values depend on are part of the compound_taskid_set.
+#'
```
```suggestion
```

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -1,17 +1,32 @@
   forecast <- hubExamples::forecast_outputs
-  forecast <- forecast[!forecast$output_type %in% c("median"), ]
+  expect_warning({
+    forecast <- forecast[!forecast$output_type %in% c("median"), ] |>
+      dplyr::mutate(output_type_id = ifelse(
+        as.numeric(output_type_id) %in% 4201:4300,
+        as.character(as.numeric(output_type_id) + 100),
+        output_type_id
+      ))
```
What are we testing here? This doesn't appear to be a test of `linear_pool` functionality.

elray1 replied:
Maybe this is just setup for the `linear_pool` tests below, and `expect_that` is being used to prevent testthat from complaining about warnings?  If so, I suggest that:
1. We add a comment explaining why we're doing this update to the `output_type_id`s
2. We find another way to avoid issues with warnings, e.g. by using `suppressWarnings` or making a more targeted update that involves casting only the `output_type_ids` that we want to change to numeric.  e.g. this could involve a check of whether `output_type_id %in% as.character(4201:4300)` to get some indices that we want to update, and then only modifying the `output_type_id` at those indices.

lshandross replied:
I added these lines of code to ensure that each unique combination of compound task ID set variables had the same output type ids to prevent an error from being thrown when this test was originally testing for something else. It's definitely more of a temporary fix than something to keep, and I meant to go file an issue on hubExamples to change the output type ids for those forecasts (especially since I don't think there's a reason not to change them)

lshandross replied:
This is no longer a blocker for merging this pull request after fixing a bug in the validation code but has none the less been filed as an issue https://github.com/hubverse-org/example-complex-forecast-hub/issues/42

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -411,6 +426,195 @@ test_that("(weighted) quantiles correctly calculated - lognormal family", {
+
+
+test_that("If the specified `compound_taskid_set` is incompatible with component model outputs, throw an error", {
+  sample_outputs <- expand.grid(stringsAsFactors = FALSE,
```
Let's add a comment describing the setup for this test.

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -411,6 +426,195 @@ test_that("(weighted) quantiles correctly calculated - lognormal family", {
+})
+
+
+test_that("Unequal samples across component models for unique of compound task ID set vars throws an error", {
```
```suggestion
test_that("Unequal samples across component models for unique combination of compound task ID set vars throws an error", {
```

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -411,6 +426,195 @@ test_that("(weighted) quantiles correctly calculated - lognormal family", {
+                         sample_outputs$output_type_id == 3] <-
+    c(150, 300, 500, 350)
+
+  sample_outputs |>
+    dplyr::mutate(horizon = 0, value = 0.75 * value) |>
+    dplyr::bind_rows(sample_outputs) |>
+    dplyr::filter(model_id %in% letters[1:3] | (output_type_id == 1 & location == "222")) |>
+    linear_pool(
+      weights = NULL,
+      task_id_cols = c("target_date", "target", "horizon", "location"),
+      compound_taskid_set = c("target", "location", "target_date"),
+      n_output_samples = NULL
+    ) |>
+    expect_error(
+      regex = "Within each group defined by a combination of the compound task ID set variables",
+      fixed = TRUE
+    )
```
This is a matter of personal preference, but I like to put the `expect_error` call at the top level rather than piped at the end.  This reads more naturally to me:
 - rather than "take the sample outputs and mutate them and add them to the existing sample outputs and then do some filtering and then compute a linear pool and then expect an error to have been thrown"
 - i read "expect an error when I call linear pool with these sample outputs"
```suggestion
 sample_outputs <- sample_outputs |>
    dplyr::mutate(horizon = 0, value = 0.75 * value) |>
    dplyr::bind_rows(sample_outputs) |>
    dplyr::filter(model_id %in% letters[1:3] | (output_type_id == 1 & location == "222"))

  expect_error(
    linear_pool(
      weights = NULL,
      task_id_cols = c("target_date", "target", "horizon", "location"),
      compound_taskid_set = c("target", "location", "target_date"),
      n_output_samples = NULL
    ),
  regex = "Within each group defined by a combination of the compound task ID set variables",
  fixed = TRUE
)
```

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -411,6 +426,195 @@ test_that("(weighted) quantiles correctly calculated - lognormal family", {
+      compound_taskid_set = c("target", "target_date"),
+      n_output_samples = NULL
+    ) |>
+    expect_error(
```
I made a comment in the next test about piping an error-generating function call into `expect_error` that also applies here.

elray1 commented on tests/testthat/test-linear_pool.R:
```diff
@@ -453,13 +657,222 @@ test_that("samples only collected and re-indexed for simplest case", {
+                         sample_outputs$output_type_id == 3] <-
+    c(150, 300, 500)
+
+  set.seed(1234)
+  models_to_resample <- sample(x = letters[1:4], size = 6 %% 4)
+  expected_outputs <- expand.grid(
+    stringsAsFactors = FALSE,
+    KEEP.OUT.ATTRS = FALSE,
+    location = c("222", "888"),
+    target = "inc death",
+    target_date = as.Date("2021-12-25"),
+    component_model = letters[1:6],
+    distinct_ids = 1
+  )
+  expected_outputs$component_model[expected_outputs$location == "222" &
+                                     expected_outputs$component_model %in% letters[5:6]] <- models_to_resample
+  expected_outputs$component_model[expected_outputs$location == "888" &
+                                     expected_outputs$component_model %in% letters[4:6]] <- letters[1:3]
+  expected_outputs <- expected_outputs |>
+    dplyr::arrange(dplyr::across(c("location", "target", "target_date", "component_model"))) |>
+    dplyr::tibble()
```
In this test and the next, I am wondering if rather than recreating the random sampling of models to create `expected_outputs`, it would make sense to test expected attributes of the actual linear pool outputs like:
 - the total number of samples adds up to the `n_output_samples`
 - within each combination of target/location/target_date,
    - for location 888, each individual model appears either 1 or 2 times
    - for location 222, models a, b, c each appear 2 times

elray1 reviewed (changes requested):
I think the actual functionality is basically good to go.  I put in pr #150 with some miscellaneous minor cleanup, mostly just adding comments.

I think all the unit tests we need are in place, but I think there's still room to do a little more tidying up.  Mostly the comments on the tests are minor/not substantive.  My only "real" comment is that for the ones about checking that samples are correctly drawn, (1) I still don't fully understand the setup with the `distinct_ids` (perhaps it would be helpful to talk through in person?) but (2) I also wonder if it would make sense to remove the actual recreation of sampling with `set.seed` and intead do higher level checks that the ensemble outputs have desired characteristics.

lshandross commented:
The following commits fix a bug I identified while investigating when derived task ids should be included in the compound task id set. When derived task ids were not included in the compound task id set, `validate_compound_taskid_set()` assumed that there is (additional) dependency across the values for those derived tasks and through a false positive error. Thus, a `derived_tasks` parameter needed to be added for that validation, even though it's not used in the `linear_pool_sample()` calculation. 

Addressing this bug also meant that I could remove the extra `mutate()` statement in the first `linear_pool()` test to fix the mismatched output type ids, which technically do not violate our requirement that sample forecasts must share output type ids across each unique combination of compound task id set variables for every model. (We may still be interested in updating the example data, though, for output type id consistency.)

lshandross commented:
The following is a summary meant to get reviewers oriented with the functionality in this pull request. Please provide your feedback by **Wednesday, January 15**.

## Context
The hubEnsembles package currently only supports the simplest case of ensembling samples using a linear pool, which adheres to the following conditions:
1. The ensemble weights all models equally
2. All component models provide an equal number of samples per compound unit
3. There is no limit on the number of samples to return (i.e., we simply pool all forecasts into an ensemble and return them)

We are interested in expanding to more complex cases, eliminating one condition at a time. *Here, we start by eliminating condition 3.* Note that we are NOT interested in eliminating conditions 1 and 2 at this time, though there is some built-in flexibility that will make eliminating them easier in the future.

## High level summary:
Adds ability to subset provided samples, implemented by a map over groups defined by the columns making up the compound task id set and sampling the output type ids of the requested number of predictions to return.

## Ensembling steps (excluding validations):
0. `linear_pool()` splits provided model inputs by output type and the helper `linear_pool_sample()` is called to act on the sample predictions
1. `weights` are validated to be equal for every model (throws an error if otherwise) OR set to be equal if not provided
2. If a subset of output samples is requested, do steps 3-4; else, skip to step 5.
3. Calculate the number of samples to output per model within each compound unit, implemented by a map. If the total number of output samples cannot be divided evenly among the models, the remainder is randomly distributed among them
  a. We use the `weights` to help determine how many samples to output for each model for a single compound unit
  b. If a model does not predict for a given compound unit, no output samples will be requested from it for that compound unit and will instead be split among the models that do predict for that compound unit
4. Draw the correct number of samples to output from each model for each compound unit (defined as a unique combination of task id set variable values) by sampling the model's output type ids, executed by a map. This grouping ensures that any samples that are joint across non-compound task id set variables will be connected and drawn together
5. Update the output type ids to be unique, so as not to falsely link any predictions that are not joint across. The data type of the `output_type_id` column is either a string or an integer, depending on the column's original data type.
6. Update the `model_id` to be that of the ensemble, which may be specified by the user

## Validation functions for ensembling samples
- `validate_compound_taskid_set()`
  - Checks that component model outputs are compatible with the specified compound task id set (i.e., samples that are joint across can be combined by their shared output type id values; note that the output type id values only need to be the same within each model)
  - Check that all component models forecast for the same set of joint across values (e.g. if samples are joint across horizons, all models must forecast for every horizon). There is an option to return a data.frame that summarizes which predictions are missing for a particular model
- `validate_sample_inputs()`
  - Checks validity of inputs to `linear_pool_sample()`
  - Checks that for each compound unit, all models provide the same number of samples

nickreich commented on R/linear_pool.R:
```diff
@@ -9,6 +9,17 @@
 #' @param ... parameters that are passed to `distfromq::make_q_fn`, specifying
 #'   details of how to estimate a quantile function from provided quantile levels
 #'   and quantile values for `output_type` `"quantile"`.
+#' @param compound_taskid_set `character` vector of the compound task ID variable
+#'   set. This argument is only relevant for `output_type` `"sample"`. NULL means
+#'   that samples are from a multivariate joint distribution across all levels of
+#'   all task id variables, while equality to `task_id_cols` means that the samples
+#'   are from separate univariate distributions for each individual prediction task.
+#'   NA means the compound_taskid_set is not relevant for the current modeling task.
+#'   Defaults to NA. Derived task ids must be included if all of the task ids their
+#'   values depend on are part of the compound_taskid_set.
```
I find this parameter definition dense and hard to wade through. I also note that no additions are made in the @details section below, where maybe they should be as all other output types are covered there. So perhaps a briefer definition here (or the same?) and then adding some additional longer-form explanations, maybe with examples and possibly links to the definitions of the concepts of compound task id and derived task id, below.

zkamvar replied:
I would suggest to list the default case first. 

Moreover, you can make this a list of the expected inputs and ROxygen will happily parse it as a list.

lshandross replied:
I've taken Zhian's suggestion of making the expected inputs a list and reordering the cases, plus added a section in the `@details` about how `linear_pool` deals with samples. Let me know if this is clearer @nickreich

nickreich commented on R/linear_pool_sample.R:
```diff
@@ -21,36 +21,63 @@ linear_pool_sample <- function(model_out_tbl, weights = NULL,
+      cli::cli_abort("Requested output samples per compound unit cannot exceed the provided samples per compound unit.")
+    }
+
+    # draw the target number of samples from each model for each unique
+    # combination of compound task ID set variables
+    split_compound_taskid_set <- model_out_tbl |>
+      split(f = model_out_tbl[, c("model_id", compound_taskid_set)])
+    model_out_tbl <- split_compound_taskid_set |>
+      purrr::map(.f = function(split_outputs) {
+        if (nrow(split_outputs) != 0) {
+          # current_compound_taskid_set has 1 row, where the column
+          # `target_samples` is the number of samples to draw for this
+          # combination of model_id and compound task ID set variables
+          current_compound_taskid_set <- split_outputs |>
+            dplyr::distinct(dplyr::across(dplyr::all_of(compound_taskid_set)), .keep_all = TRUE) |>
+            dplyr::left_join(samples_per_combo, by = c("model_id", compound_taskid_set))
+          provided_indices <- unique(split_outputs$output_type_id)
+
+          sample_idx <- sample(x = provided_indices, size = current_compound_taskid_set$target_samples, replace = FALSE)
+          dplyr::filter(split_outputs, .data[["output_type_id"]] %in% sample_idx)
+        }
+      }) |>
+      purrr::list_rbind()
   }
```
I am not sure if the goal should be that the code is "readable" but I spent 10 minutes trying to parse this code (without running it) and can't understand what it is doing. Perhaps some more comments would help. 

specifically: 
1. I don't understand what the split_compound_taskid_set object is, even after looking at the base::split() helpfile.
2. I don't understand what the nested set of dplyr functions on line 72 is doing.

lshandross replied:
@zkamvar I have a couple of questions about points (1) and (2):
1. For the refactoring into a separate function, which lines should be included in the refactor? Everything from 61 to 80? A subset of that?
2. For the control structure, are you suggesting that I write two separate functions for the different cases? Or just the `do_this()` case?  Also, is this refactoring supposed to be only for within the map, or for more which includes the map?

zkamvar replied:
> 1. For the refactoring into a separate function, which lines should be included in the refactor? Everything from 61 to 80? A subset of that?

The anonymous function that you have from lines 66--79. Inline functions are generally okay, but once they get longer than 5 lines, they really should stand alone otherwise it's difficult to debug them down the road.

> 2. For the control structure, are you suggesting that I write two separate functions for the different cases? Or just the `do_this()` case?  Also, is this refactoring supposed to be only for within the map, or for more which includes the map?

I'm so sorry, think I over-engineered my answer 😞. The thing I'm describing here is called [a guard clause](https://arne-mertz.de/2017/08/code-smells-short-list/#Deeply_nested_control_flow).  

Your function currently consists of an if statement that allows the function to do something. When that statement is FALSE, then the function does nothing.    Instead, if you add a guard clause at the beginning that returns early if the function shouldn't do anything to the input, then it helps reduce indentation levels, which makes things easier to read and debug. 

```r
# good
  if (nrow(thing) == 0) {
    return(thing)
  }
  # do stuff

# -----------------------

# bad
  if (nrow(thing) > 0) {
    # do stuff
  }
```

Let me know if you want to do some pairing on this.

lshandross replied:
Thanks for the clarifications! I just wanted to make sure I was fully understanding everything

nickreich reviewed (approved):
I have two minor, non-essential comments. I read through the code carefully, but did not run anything on my own. Basically, I think the changes look good, and the strategy of tackling one assumption at a time also seems reasonable.

zkamvar reviewed (approved):
Good work! I appreciated the comments that provided context and I really appreciated the cleanup of `test-linear_pool.R` by adding `helper-test_data.R`, it makes it clear what is being tested 🙌🏽 

I will admit that I don't fully grok how the compound task ID set is used. I get the concept of setting `NULL` to mean that samples are derived from joint distribution across all of the task IDs, but when it comes to `NA` or specifying a set that matches the task ID cols... I get kinda lost. It's not your fault, the sample thing is inherently confusing. 

As with Nick, I had minor comments that would help improve the maintainability of the code.

And again, good work!

elray1 commented on R/validate_ensemble_inputs.R:
```diff
@@ -168,9 +171,12 @@ validate_compound_taskid_set <- function(model_out_tbl,
   }
 
-  dependent_tasks <- setdiff(task_id_cols, compound_taskid_set)
+  dependent_tasks <- setdiff(task_id_cols, c(compound_taskid_set, derived_tasks))
```
I think the logic here and below is correct, I was just confused by the name because some of the things in `derived_tasks` may be task id variables for which a model's samples capture dependence. Adding a comment here just to orient the reader to what this is intended to include.
```suggestion
  # the set of "non-derived" task id variables for which the samples should capture dependence
  dependent_tasks <- setdiff(task_id_cols, c(compound_taskid_set, derived_tasks))
```

lshandross replied:
I've taken the suggestion, but do you also think it would be good to rename this variable? Also any suggestions for a different name?

lshandross commented on tests/testthat/test-linear_pool.R:
```diff
@@ -411,38 +366,148 @@ test_that("(weighted) quantiles correctly calculated - lognormal family", {
+  # test that df of missing combos returns the expected value
+  missing_expected <- create_test_sample_outputs() |>
+    dplyr::filter(!(model_id %in% letters[1:3] | horizon == 1)) |>
+    dplyr::distinct(dplyr::across(dplyr::all_of(c("model_id", setdiff(sample_tasks, NULL)))))
```
I added the `NULL` to demonstrate that the code is the same as the block below, but this case has no derived tasks. I've updated inline comments to better reflect this

zkamvar reviewed (approved):
LGTM! Thank you, @lshandross! I appreciate you taking the time to refactor the inline functions and to add documentation. 

I've added some suggestions for nice-to-haves and bringing your attention to the ordering of output type IDs, but I don't think any of them are dealbreakers.

nickreich reviewed (approved):
Thanks for all the changes @lshandross! I approve of the updates.
